### DISCUSSION TITLE
Multiple prover backend design (Z3, Singular, Z3 with different options)

### DISCUSSION THREAD

**Discussion by chanheec** [verus-lang/verus#99] [General]
Reasons for multiple prover backend are:
1. Bit-vector reasoning works better with different options from the default Verus settings. (related to this issue: https://github.com/secure-foundations/verus/issues/49#issue-1119725461)
2. There is a plan to add nonlinear arithmetic support using Singular (porting the work of @yizhou7 in Dafny to Verus)
3. Also, we might want a fresh Z3 to clear up the context sometimes.

I would like to know what would be a reasonable way to do this.

**Comment by chanheec** [comment]
From today's discussion:

The current plan is to make `Z3 restart primitive` first - making a new Z3 process and loading default options and prelude (and nothing more)
After having this, we can use this for every bit-vector query. (If this approach turns out to be inefficient, we can use a long-living bit-vector dedicated Z3 process) (We might consider producing some Z3 processes in advance, and make use of it later)
 
For non-linear arithmetic, we can spawn a long-living Singular process and use it. If Singular takes a negligible amount of time to restart, we might restart it every time we use it.
For users without Singular, I will try to separate test cases that can be run in Z3 and those only work with Singular.

**Comment by utaal** [comment]
Most of this has now been implemented by @channy412.
